src/go/rpk/pkg/cli/cluster/partitions/move.go

```diff
@@ -0,0 +1,283 @@
+				partition := partition
+				g.Go(func() (rerr error) {
+					if len(topics) > 0 { // foo -p 0:1,2,3
+						for _, t := range topics {
+							_, _, part, err := extractNTP(t, partition)
+							out.MaybeDie(err, "failed to extract topic/partition: %s\n", err)
+
+							if nt := strings.Split(t, "/"); len(nt) == 1 {
+								ns = "kafka"
+								topic = nt[0]
+							} else {
+								ns = nt[0]
+								topic = nt[1]
+							}
+
+							current, err := cl.GetPartition(egCtx, ns, topic, part)
+							out.MaybeDie(err, "unable to get partition: %s\n", err)
+
+							newReplicas, err := configureReplicas(partition, current.Replicas)
+							out.MaybeDie(err, "unable to configure new replicas: %v\n", err)
+
+							newAssignmentList = append(newAssignmentList, newAssignment{
+								Namespace:   ns,
+								Topic:       topic,
+								Partition:   part,
+								NewReplicas: newReplicas,
+							})
+						}
+					} else { // -p foo/0:1,2,3
+						ns, topic, part, err := extractNTP("", partition)
+						out.MaybeDie(err, "failed to extract topic/partition: %s\n", err)
+
+						current, err := cl.GetPartition(egCtx, ns, topic, part)
+						out.MaybeDie(err, "unable to get partition: %s\n", err)
+
+						newReplicas, err := configureReplicas(partition, current.Replicas)
+						out.MaybeDie(err, "unable to configure new replicas: %v\n", err)
+
+						newAssignmentList = append(newAssignmentList, newAssignment{
+							Namespace:   ns,
+							Topic:       topic,
+							Partition:   part,
+							NewReplicas: newReplicas,
+						})
+					}
+					return nil
+				})
+			}
+
+			if err := g.Wait(); err != nil {
+				fmt.Printf("failed to begin movements: %v\n", err)
+			}
+
+			knownNodeCore := make(map[int]int)
+			var reqCount int
+
+			// We do call MoveReplicas serially to ensure each iteration
+			// refers to the updated 'knownNodeCore'
+			for i, newa := range newAssignmentList {
+				for j, nr := range newa.NewReplicas {
+					if nr.Core == -1 {
+						numCore, exists := knownNodeCore[nr.NodeID]
+						if exists {
+							src := rand.NewSource(time.Now().UnixNano())
+							r := rand.New(src)
+							newAssignmentList[i].NewReplicas[j].Core = r.Intn(numCore)
+						} else {
+							broker, err := cl.Broker(cmd.Context(), nr.NodeID)
+							out.MaybeDie(err, "unable to find the current core assignments", err)
+							src := rand.NewSource(time.Now().UnixNano())
+							r := rand.New(src)
+							newAssignmentList[i].NewReplicas[j].Core = r.Intn(broker.NumCores)
+							knownNodeCore[nr.NodeID] = broker.NumCores
+						}
+					}
+				}
+				fmt.Printf("Requested to move %s/%s/%d to %s\n", newa.Namespace, newa.Topic, newa.Partition, formatNodeCore(newa.NewReplicas))
+				reqCount += 1
+				err = cl.MoveReplicas(cmd.Context(), newa.Namespace, newa.Topic, newa.Partition, newa.NewReplicas)
+				if err != nil {
+					fmt.Printf("unable to move replicas for partition %s/%s/%d: %v\n", newa.Namespace, newa.Topic, newa.Partition, err)
```

Comment:
This is what I mean / what I recommend switching to:
```go
var mu sync.Mutex
knownNodeCore := make(map[int]int)
for i, newa := range newAssignmentList {
	brokerReqs := make(map[int32]struct{})
	for j, nr := range newa.NewReplicas {
		if nr.Core == -1 {
			brokerReqs[nr.NodeID] = struct{}{}
		}
	}
	for node := range brokerReqs {
		node := node
		g.Go(func() error) {
			broker, err := cl.Broker(cmd.Context(), node)
			mu.Lock()
			defer mu.Unlock()
			knownNodeCore[nr.NodeID] = broker.NumCores
			return err
		})
	}
}
if err := g.Wait(); err != nil {
	out.MaybeDie(err, "unable to find the current core assignments", err)
}


rng := rand.New(rand.NewSource(time.Now().UnixNano()))
for i := range newAssignmentList {
	newa := newAssignmentList[i]
	for j, nr := range newa.NewReplicas {
		if nr.Core == -1 {
			numCore := knownNodeCore[nr.NodeID]
			newa.NewReplicas[j].Core = rng.Intn(numCore)
		}
	}
	g.Go(func() error) {
		return cl.MoveReplicas(cmd.Context(), newa.Namespace, newa.Topic, newa.Partition, newa.NewReplicas)
	})
}
if err := g.Wait(); err != nil {
	fmt.Printf("unable to move replicas for partitions: %v", err)
}
```

This allows each series of requests to make to be concurrent

Reply:
Thanks. Well, in order to call `Broker` just once per a node, I'd make this loop `for node := range brokerReqs {` an independent one like this
```Go
			knownNodeCore := make(map[int]int)
			brokerReqs := make(map[int]struct{})
			for _, newa := range newAssignmentList {
				for _, nr := range newa.NewReplicas {
					if nr.Core == -1 {
						brokerReqs[nr.NodeID] = struct{}{}
					}
				}
			}
			for node := range brokerReqs {
				node := node
				g.Go(func() error {
					broker, err := cl.Broker(cmd.Context(), node)
					mu.Lock()
					defer mu.Unlock()
					knownNodeCore[node] = broker.NumCores
					return err
				})
			}
```
However, UX-wise, I have some concerns below with errgroup:

* making everything fail when one of the requests fails doesn't make sense. A failure of a request shouldn't affect the other request as they're independent.
* capturing which request succeeds and which one fails correctly. With errgroup, we can capture only one failure even there are multiple errors at play, right?

Reply:
I see, I'll modify it a bit -- to be honest, I don't personally use errgroup and just use goroutines directly with locks.

Reply:
Indeed it makes it more controllable.